### ISSUE TITLE
[BUG] Xamarin : Broken Bitmap.Encode on real devices

### ISSUE THREAD

**Issue by NovaKs68** [mono/SkiaSharp#2643]
### Description

The image is darkened when encoded by SkiaSharp.
This behavior occurs on physicals devices, but not on :
- Android emulator
- iOS emulator

So I want to be able to Decode an image and Encode it with SkiaSharp without changing the color. 

**Version of Skiasharp:** 2.88.6
**Last Known Good Version of SkiaSharp:** 2.88.5

### Code

Link to a reproduction project repository: https://github.com/NovaKs68/bug-skiasharp

The following two code blocks modify the image color:
```cs
using (var stream = File.OpenRead(imagePath))
{
    SKImage img = SKImage.FromEncodedData(stream);
    SKBitmap bitmap = SKBitmap.FromImage(img);

    using (MemoryStream memStream = new MemoryStream())
    using (SKManagedWStream wstream = new SKManagedWStream(memStream))
    {
        bitmap.Encode(wstream, SKEncodedImageFormat.Jpeg, 100);
        imageCompressedView.Source = ImageSource.FromStream(() => new MemoryStream(memStream.ToArray()));
    }
}
```

```cs
var codec = SKCodec.Create(imagePath);
var skImage = SKBitmap.Decode(codec);

using (MemoryStream memStream = new MemoryStream())
using (SKManagedWStream wstream = new SKManagedWStream(memStream))
{
    skImage.Encode(wstream, SKEncodedImageFormat.Jpeg, 100);
    imageCompressedView.Source = ImageSource.FromStream(() => new MemoryStream(memStream.ToArray()));
}
```

### Expected Behavior

The result image should be similar to the input image.

### Actual Behavior

The result image is darker than the input image (see screenshots).

### Version of SkiaSharp

Other (Please indicate in the description)

### Last Known Good Version of SkiaSharp

Other (Please indicate in the description)

### IDE / Editor

Visual Studio (Windows)

### Platform / Operating System

Android, iOS

### Platform / Operating System Version

- Android 11
- iOS 17.0.3

### Devices

- Xiaomi Redmi 11 Pro
- IPhone 12 Pro Max

### Relevant Screenshots

On iOS:
![MicrosoftTeams-image (18)](https://github.com/mono/SkiaSharp/assets/55892105/7791d69d-0077-41f7-8f03-4fd918aa0d59)

On Android:
![2023-10-10_11-49-33](https://github.com/mono/SkiaSharp/assets/55892105/6bb33e39-5dda-44e7-8e4d-2e43cad19e5f)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by espenrl** [comment]
I experienced the exact same thing and had to roll back to a previous version.

**Comment by happyfr34k** [comment]
It's exactly the same for us, we did a downgrade to the previous version but since there's a high-severity vulnerability (https://github.com/advisories/GHSA-j7hp-h8jx-5ppr), it would be really nice to have a patch for that...

**Comment by jhariel8** [comment]
Echoing what everyone else has said, we are experiencing a similar issue on our Xamarin.iOS project, except rather than just being "darker," the images seem to either be completely black or look really corrupt. Just a couple examples of what some photos look like:

![0991f6c7-d588-471b-8b8f-0b03c1940a03](https://github.com/mono/SkiaSharp/assets/23176078/8365cc76-5b32-4807-bbfd-6b8f5a2cfb54)
![a7512b50-835b-4d9f-8edf-e2c367194879](https://github.com/mono/SkiaSharp/assets/23176078/cebf5071-d574-4795-b71e-2e18a94b3131)

Photos look normal when running on UWP and in the simulator, but when deployed to a physical iOS device they are corrupted like this. Downgrading our Xamarin.iOS project to SkiaSharp version 2.8.8.3 fixed this issue for us. We are going with this solution for now because this concerns a critical part of our application, but since there is a major vulnerability in this package, a fix is very much needed.

**Comment by jonmdev** [comment]
Oh wow, I just spent the last hour pulling my hair out over this same issue trying to figure out if I was doing something wrong. Glad to know I'm not the only one. I only get this bug in iOS.

Thanks for stopping me from Googling endlessly searching for solutions. 👍 

Anyone who can, please fix.

**Comment by IainS1986** [comment]
Just echoing I comment I made on another, related issue.

I'm seeing similar issues but manifesting in PDF documents produced via Skiasharp.

I'm still investigating but my hunch is that it's JPG encoding specifically. Which is a pain for myself as there's no alternative as I think PDF encoding takes all image data and converts to JPG.

However - I'm not seeing **all** images come out corrupt. I have one test where one image is fine and one isn't. The difference is alpha channel.

https://github.com/mono/SkiaSharp/issues/2650#issuecomment-1781968854

Be interesting to know if any of you seeing issues - if you can - can you change one of your images to RGBA if it's only RGB and does that produce any difference? Or can any of you share what colorspace/encoding format your images are in?

**Comment by NovaKs68** [comment]
Hello @IainS1986 , unfortunately the images uploaded to the test project are in JPG and RGBA format.

**Comment by IainS1986** [comment]
> Hello @IainS1986 , unfortunately the images uploaded to the test project are in JPG and RGBA format.

JPGs don't support transparency, so they will be in RGB format.

Might confirm a bit more that it's JPG/RGB images specifically...

**Comment by beeradmoore** [comment]
Also having this problem after upgrading to 2.88.6. Parts of our code is in a horrible state of "This is awful, but if we touch anything it will likely fall apart. So we best wait until MAUI port and re-write those key bits completely". We 100% have places where files are saving/opening/copying/renaming/etc and we have jpg data in .png files and png data in .jpg files (and I have hated it ever since I discovered it).

Getting both corruption blocks like what @jhariel8 showed, with some tinting that @NovaKs68 showed. Sometimes even getting images shift horizontally ~40% with the other image data looping back. 

Additionally I think the sample repo is doing something bad (see [this line](https://github.com/NovaKs68/bug-skiasharp/blob/main/TestProject/TestProject/Views/AboutPage.xaml.cs#L29)). If a user selects a png from their gallery it will have its data copied and saved with .jpeg extension. I'd imagine that wouldn't cause issues because `SKImage.FromEncodedData(` is unaware of the filename and instead use the headers. This is similar to the path we went down to do many many mistakes with images where everything is everywhere and its all bad 😅

**Comment by KRA2008** [comment]
i'm encountering this issue too. i have an open source project that i can change the SkiaSharp versions ing (and nothing else) and make the issue appear and disappear. if more information about recreating the issue in my project is needed i can provide it. 

i upgraded from 2.80.3 to 2.88.6 to fix issue #2265, but then i encountered this issue and i tried to work around it by downgrading to 2.88.5, but then i get blocked by #2584. i'll just keep downgrading.

**Comment by IainS1986** [comment]
Looking at the change log from 2.88.5 to 2.88.6 there isn't any actual logic changes made.

Various dependencies updated to newer versions, with a suspect one settings libjpeg being removed and libjpeg-turbo being updated a considerable amount of versions.

Wonder if we can at least have those rolled back if that's the cause to get a new build of Skia without the vulnerabilities in it?

**Comment by IainS1986** [comment]
> i upgraded from 2.80.3 to 2.88.6 to fix issue #2265, but then i encountered this issue and i tried to work around it by downgrading to 2.88.5, but then i get blocked by #2584. i'll just keep downgrading.

Yes, if you're deploying to iOS you have to downgrade to 2.88.3.

**Comment by MFinkBK** [comment]
@mattleibow Can you say something about this bug? We had to roll back to SkiaSharp 2.88.3 since this is a regression from 2.88.5, but 2.88.5 and 2.88.4 doesn't allow publishing in Apple App Store, and so we're stuck on an old version with the CVE-2023-4863 vulnerability fixed in #2608.

**Comment by MarcinZiabek** [comment]
I would like to confirm the problem. It is related to encoding/decoding of JPEG images only. Based on my testing, it happens on macOS arm64 (M2). 

Link to repository that can be used for bug reproduction: https://github.com/MarcinZiabek/SkiaSharp.JpegEncodeIssue

**Comment by IainS1986** [comment]
@MarcinZiabek 

Can you check png images with no alpha channel, only RGB (not png images with opaque alpha, specifically RGB only png images) in your repo?

I'm pretty certain I've seen issues with those as well, but have no repo to show it


My hunch is it's not JPG specific but RGB data being wrongly encoded into RGBA (and so all JPGs get affected)

**Comment by MarcinZiabek** [comment]
@IainS1986 The following code generates a PNG image with a specified color type. The gradient is clipped to ensure transparency configuration.

```csharp
void Generate(SKColorType colorType)
{
    var imageInfo = new SKImageInfo(400, 300, colorType);
    using var surface = SKSurface.Create(imageInfo);

    surface.Canvas.ClipRect(new SKRect(50, 50, 300, 200));
        
    using var shaderPaint = new SKPaint
    {
        Shader = SKShader.CreateRadialGradient(
            new SKPoint(200, 150), 300,
            new[] { SKColors.DeepSkyBlue, SKColors.Green }, new[] {0, 1f},
            SKShaderTileMode.Decal)
    };
            
    surface.Canvas.DrawRect(0, 0, imageInfo.Width, imageInfo.Height, shaderPaint);

            
// return result as an image
    using var image = surface.Snapshot();
    using var data = image.Encode(SKEncodedImageFormat.Png, 90);
    File.WriteAllBytes($"result.png", data.ToArray());
}
```

1. Opaque:

```csharp
Generate(SKColorType.Rgb565);
```

![result_opaque](https://github.com/mono/SkiaSharp/assets/9263853/29e56ceb-cd1c-4b36-bf8c-b49f6e62dff6)

2. With transparency channel:

```csharp
Generate(SKColorType.Rgba8888);
```

![result_opaque](https://github.com/mono/SkiaSharp/assets/9263853/9fa2d0a8-5fee-4b95-a5b5-0a919ffbab24)


Tested on: MacBook Pro 16 M2 Max ARM64, Sonoma 14.1.1

**Comment by MarcinZiabek** [comment]
From my observations:
- loading PNG image and saving as JPEG: ❌ broken
- loading JPEG image and saving as PNG: ✅ working
- generating image and saving as JPEG: ❌ broken
- generating image and saving as PNG: ✅ working

Tested on: MacBook Pro 16 M2 Max ARM64, Sonoma 14.1.1

**Comment by IainS1986** [comment]
Hmm fair enough, I was meaning more with decoding then encoding external images, not sure if a test using Skia to generate the PNG itself in the first place would result in the same or different issues if it's image encoding that seems broken? We have to trust the original generation 'worked'?

There's also a chance skia generates pngs with no alpha fine, but fails in decoding the raw byte data of the same image from external sources?

I definitely see issues with PDF document generation though which I believe concerts all images to JPG so I'm 100% seeing the same JPG issue - I'm just sure I saw an issue with an RGB PNG in a general SKCanvas view (on Android) that showed similar corruption 🤷‍♂️

**Comment by SeriousM** [comment]
I can contribute to the issue:
we're using 2.88.6

Original image:
![original image](https://github.com/mono/SkiaSharp/assets/462013/e0b49c84-88b5-4e2a-bdfd-f14a5741cadf)

Encode on windows (works):
![win-full-size-only-redrawn](https://github.com/mono/SkiaSharp/assets/462013/67e8a8ba-bad3-4eb2-8eba-b9fb09bb8e3e)

Encode on android (same result on ios; broken):
![andr-full-size-only-redrawn](https://github.com/mono/SkiaSharp/assets/462013/0ba5028e-5e73-46ba-bca5-33ae1b58a51d)

Using https://www.guiffy.com/Binary-Diff-Tool.html I found a difference in a small amount of bytes:
![diff](https://github.com/mono/SkiaSharp/assets/462013/3ecfc5e9-c374-4523-891c-c8ac446f96c8)

To my analysis, the broken image is shifted to the right and having an artifact at the top left. Sometimes the image is darkened, brightened, pink, purple, cyan... just a broken calculation of the bits based on the bytes different.
![image](https://github.com/mono/SkiaSharp/assets/462013/ed10ab6f-f1f1-4df3-a74e-dc491b7d17b4)


That's the code I used to resize the image:
```cs
  private static Stream resizeImage(Stream sourceImageStream, SKEncodedImageFormat imageFormat, int quality)
  {
    // make sure the image stream is readable from the start
    try { sourceImageStream.Position = 0; } catch (NotSupportedException) { }

    using var inputStream = new SKManagedStream(sourceImageStream);
    using var original = SKImage.FromEncodedData(inputStream);

    var imageWidth = original.Width;
    var imageHeight = original.Height;

    var useWidth = imageWidth;
    var useHeight = imageHeight;

    var info = new SKImageInfo(useWidth, useHeight, original.ColorType, original.AlphaType, original.ColorSpace);
    using var surface = SKSurface.Create(info);
    surface.Canvas.Clear();
	
    surface.Canvas.DrawImage(original,
      SKRect.Create(0, 0, useWidth, useHeight)
    );
    surface.Canvas.Flush();

    using var snapshot = surface.Snapshot();

    using var data = snapshot.Encode(imageFormat, quality);
    var resizedContent = data.ToArray();

    var resizedContentStream = new MemoryStream(resizedContent);

    return resizedContentStream;
  }
```

I'm sure the code could be shorter (as others posted) but I wanted to give you a way to reproduce the exact same bytes.

I really hope that helps!

**Comment by ispysoftware** [comment]
Also getting jpeg encode issues on Mac OS only (Android / Windows OK) 

288.5:
![image](https://github.com/mono/SkiaSharp/assets/800093/8a12792e-eeb5-483d-89d8-9c7fb8b2d234)

288.6:
![image](https://github.com/mono/SkiaSharp/assets/800093/843ff4ce-6671-4a97-a8c8-fd71fb8feea8)

**Comment by Drogo0716** [comment]
Can confirm this happened for my organization as well with our Android application. Upgraded to SkiSharp 2.88.6 due to a security vulnerability. Each image decoded to a SKBitmap while using 2.88.6 gets corrupted.

**Comment by r-work** [comment]
Same here issue, on MAUI Android.

**Comment by jasoncodematic** [comment]
Seems like there have been no releases/commits since Sep-2023 (approx 3 months at the time of writing this) and this above critical issue has been raised a number of times. Does anyone know if this repository is abandoned?

**Comment by IainS1986** [comment]
With the release of .net-8.0, which is meant to be the saving graces for MAUI - and SkiaSharp is pretty closely linked. Maybe the .net-8 release is just taking major priority over everything else?

It is a little concerning...

**Comment by simotesttest** [comment]
I experienced the exact same thing in Android and iOS

**Comment by ispysoftware** [comment]
@mattleibow  i'm not sure what your involvement is in this but if you could take a look at this critical issue i'm sure we'd all appreciate it!

**Comment by orwo1** [comment]
I'm hitting the same issue with 2.88.6, can't rollback cause of other nugets dependent on that.
Reproduced in Maui app, both Android and iOS.
The problematic code:

    `public static class ResizeImageHelper
    {
    // 320 - because we want to limit image size to 350KB
    private const int ImageWidthAndHeight = 320;
    private const int Quality = 100;

    public static byte[] ResizeImageSource(Stream stream)
    {
        try
        {
            byte[] imageAsByte = stream.ToByteArray();

            var bitmap = SKBitmap.Decode(imageAsByte);
            var scaled = bitmap.Resize(new SKImageInfo(ImageWidthAndHeight, ImageWidthAndHeight), SKFilterQuality.High);
            SKImage image = SKImage.FromBitmap(scaled);
            SKData jpgData = image.Encode(SKEncodedImageFormat.Jpeg, Quality);

            byte[] resizedImageAsByte = jpgData.ToArray();
            return resizedImageAsByte;
        }
        catch
        {
            return null;
        }
    }`

**Comment by mjsb212** [comment]
Any fix or Workaround? - Having same problem with wrong colors / distorted results. On Maui Android .net 8.0.3

**Comment by stoff99** [comment]
Same problem here with 2.88.6. Working with 2.88.5 but useable for us because of the vulnerability.

**Comment by mattleibow** [comment]
I tested 3.0.1 and the latest main and I think this is an issue in libjpeg-turbo. For now I am reverting to the latest 2.x series.

I had a look at the original security issue and it does not affect the 2.x series because 12-bit jpegs are not actually supported in our build. The 2.x series required the entire library to be either 8-bit or 12-bit, and we use 8-bit. The 3.x series allowed for both simultaneously.

The security issue was just reported for 2.1.90 (which seems like a 2.x series but is actually a 3.0 beta). The current stable version is 2.1.5.1 which is on a separate branch.

I have opened an issue in libjpeg-turbo where I will try and determine the real cause and where this issue should live: https://github.com/libjpeg-turbo/libjpeg-turbo/issues/741
